WMF Framework/Event Platform Client/EventPlatformClientLibrary.swift

```diff
@@ -0,0 +1,776 @@
+        return result
+    }
+
+    /**
+     * Submit an event according to the given stream's configuration.
+     * - Parameters:
+     *      - stream: Name of the stream to submit the event to
+     *      - data: A dictionary of event data, with appropriate `$schema`
+     *      - domain: Optional domain to include for the event (without protocol)
+     *
+     * Regarding `$schema`: the instrumentation needs to specify which schema (and
+     * specifically which version of that schema) it conforms to. Analytics
+     * schemas can be found in the jsonschema directory of
+     * [secondary repo](https://gerrit.wikimedia.org/g/schemas/event/secondary/)
+     *
+     * As an example, if instrumenting client-side error logging, a possible
+     * `$schema` would be `/mediawiki/client/error/1.0.0`. For the most part, the
+     * `$schema` will start with `/analytics`, since there's where
+     * analytics-related schemas are collected. An example call:
+     *
+     * ```
+     * EPC.shared?.submit( "test.instrumentation", [ "$schema": "/analytics/test/1.0.0" as NSCoding ] )
+     * ```
+     *
+     * Regarding `domain`: this is *optional* and should be used when event needs
+     * to be attrributed to a particular wiki (Wikidata, Wikimedia Commons, a
+     * specific edition of Wikipedia, etc.). If the language is NOT relevant in
+     * the context, `domain` can be safely omitted. Using "domain" rather than
+     * "language" is consistent with the other platforms and allows for the
+     * possibility of setting a non-Wikipedia domain like "commons.wikimedia.org"
+     * and "wikidata.org" for multimedia/metadata-related in-app analytics.
+     * Instrumentation code should use the `host` property of a `URL` as the value
+     * for this parameter.
+     *
+     * Cases where instrumentation would set a `domain`:
+     * - reading or editing an article
+     * - managing watchlist
+     * - interacting with feed
+     * - searching
+     *
+     * Cases where it might not be necessary for the instrument to set a `domain`:
+     * - changing settings
+     * - managing reading lists
+     * - navigating map of nearby articles
+     * - multi-lingual features like Suggested Edits
+     * - marking session start/end; in which case schema and `data` should have a
+     *   `languages` field where user's list of languages can be stored, although
+     *   it might make sense to set it to the domain associated with the user's
+     *   1st preferred language – in which case use
+     *   `MWKLanguageLinkController.sharedInstance().appLanguage.siteURL().host`
+     */
+    @objc public func submit(stream: String, data: [String: NSCoding], domain: String? = nil) -> Void {
+        guard self.sharingUsageData else {
+            return
+        }
+
+        /*
+         * EventGate needs to know which version of the schema to validate against
+         */
+        let schemaKey = "$schema"
```

Comment:
I put it in `schemaKey` originally, at the time I think the same literal was used in multiple places so I wanted to avoid a situation where it accidentally is only changed in one place in the future. Agree it can be moved back to a literal now though.

Reply:
Hah! Our operating theory was that it was for performance, micro-optimization reasons. Alright, I'll change this and the other `{field}Key`s too.